### PR TITLE
[Fabric] Gesture Handler for Fabric

### PR DESCRIPTION
**Pull request by tomekzaw**
## Description

This PR adds Fabric support to React Native Gesture Handler.

Co-authored by @j-piasecki.

### Main changes

* View Flattening with Fabric enabled (in some cases it may be necessary to add `collapsable={false}` in client apps)
* Old API & direct vs. device events (see [details](https://github.com/software-mansion/react-native-gesture-handler/pull/1893#discussion_r814109074))

### Building FabricExample app in release mode

<details>

The following errors may occur when building FabricExample app in release mode:

1. `Invalid regular expression: Quantifier has nothing to repeat`

Hermes does not support optional named groups in regular expressions but `ansi-styles` library uses them:

```js
const matches = /(?<colorString>[a-f\d]{6}|[a-f\d]{3})/i.exec(hex.toString(16));
```

As a workaround, you can comment out the implementation directly in `node_modules`.

2. `Undefined symbol: folly::SharedMutexImpl<false, void, std::__1::atomic, false, false, false>::deferredReaders`

As a workaround, you can comment out `use_flipper!()` in your Podfile.

</details>

### CODE REVIEW

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerButtonViewManager.kt [hunk 1]
```diff
@@ -17,32 +17,39 @@ import android.view.View.OnClickListener
 import android.view.ViewGroup
 import androidx.core.view.children
 import com.facebook.react.bridge.SoftAssertions
-import com.facebook.react.uimanager.PixelUtil
-import com.facebook.react.uimanager.ThemedReactContext
-import com.facebook.react.uimanager.ViewGroupManager
-import com.facebook.react.uimanager.ViewProps
+import com.facebook.react.module.annotations.ReactModule
+import com.facebook.react.uimanager.*
```
Comment (Ubax) [inline android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerButtonViewManager.kt:21]: Do you want to have `*` import here?

File: src/fabric/RNGestureHandlerRootView.tsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+export default function GestureHandlerRootView({
+  children,
+  ...rest
+}: GestureHandlerRootViewProps) {
+  return (
+    <GestureHandlerRootViewNativeComponent {...rest}>
+      {children}
+    </GestureHandlerRootViewNativeComponent>
+  );
+}
```
Comment (Ubax) [inline src/fabric/RNGestureHandlerRootView.tsx:20]: I think that `<GestureHandlerRootViewNativeComponent {...rest}>{childres}</...>` is the same as `<GestureHandlerRootViewNativeComponent {...props} />`. 

Is there a specific reason why you divide the props here?

**Review by Ubax — COMMENTED** [review]
I added some comments. I didn't check the GH specific logic though

File: FabricExample/android/settings.gradle [hunk 3]
```diff
@@ -1,9 +1,13 @@
 rootProject.name = 'FabricExample'
 apply from: file("../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesSettingsGradle(settings)
 include ':app'
-includeBuild('../node_modules/react-native-gradle-plugin')
 
 if (settings.hasProperty("newArchEnabled") && settings.newArchEnabled == "true") {
+    includeBuild('../node_modules/react-native-gradle-plugin') {
+        dependencySubstitution {
+            substitute(module("com.facebook.react:react")).using(project(":"))
```
[the remark this answers is not in the document]
Comment (tomekzaw) [reply FabricExample/android/settings.gradle:8]: Basically it's the same piece of code as in [react-native-screens](https://github.com/software-mansion/react-native-screens/blob/7c6e14b45834cf009c7069d46a528cf792d2a15e/FabricExample/android/settings.gradle#L4-L8) but moved inside `if`.

[the remark this answers is not in the document]
Comment (Ubax) [reply FabricExample/android/settings.gradle:8]: ok

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerButtonViewManager.kt (hunk 1 above)
Comment (j-piasecki) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerButtonViewManager.kt]: I changed it to specific imports in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/89232814dc7aa64bda39cb1874016771a99ebfa0. There was a problem where I couldn't import a specific file but `*` made it work, it seems to be ok now.

File: src/fabric/RNGestureHandlerRootView.tsx (hunk 2 above)
Comment (j-piasecki) [reply src/fabric/RNGestureHandlerRootView.tsx]: It seems like the `fabric/...` wrappers weren't necessary at all. But your comment also applies to the older `GestureHandlerRootView.android.tsx`. I don't think there is a reason for it to be this way. I changed it to match `GestureHandlerRootView.tsx` in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/50f054f59637198cecad15875547887d0a6ac2e4.

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerButtonViewManager.kt [hunk 4]
```diff
@@ -333,10 +348,10 @@ class RNGestureHandlerButtonViewManager : ViewGroupManager<ButtonViewGroup>() {
         SoftAssertions.assertNotNull(attr)
         return when (attr) {
           SELECTABLE_ITEM_BACKGROUND -> {
-            R.attr.selectableItemBackground
+            android.R.attr.selectableItemBackground
```
Comment (WoLewicki) [inline android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerButtonViewManager.kt:351]: why this is needed? `android.R` seems to be a bit different from `R`

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt [hunk 5]
```diff
@@ -518,83 +542,122 @@ class RNGestureHandlerModule(reactContext: ReactApplicationContext?)
     handlerFactories.firstOrNull { it.type == handler.javaClass } as HandlerFactory<T>?
 
   private fun <T : GestureHandler<T>> onHandlerUpdate(handler: T, motionEvent: MotionEvent) {
+    // onUpdate
```
Comment (WoLewicki) [inline android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt:545]: what does this comment concern?

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt [hunk 6]
```diff
@@ -518,83 +542,122 @@ class RNGestureHandlerModule(reactContext: ReactApplicationContext?)
...
+
     if (handler.tag < 0) {
       // root containers use negative tags, we don't need to dispatch events for them to the JS
       return
     }
     if (handler.state == GestureHandler.STATE_ACTIVE) {
       val handlerFactory = findFactoryForHandler(handler)
 
-      if (handler.usesDeviceEvents) {
+      if (handler.actionType == GestureHandler.ACTION_TYPE_REANIMATED_WORKLET) {
```
Comment (WoLewicki) [inline android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt:554]: All those changes are needed due to different handling of events in Fabric? If so, maybe it would be good to give a quick explanation why is it like that now for other libraries maintainers, or is it something that will happen only in this library?

File: android/src/main/jni/RNGestureHandlerComponentsRegistry.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  delegate->buildRegistryFunction = buildRegistryFunction;
+  return instance;
+}
+
+void RNGestureHandlerComponentsRegistry::registerNatives() {
+  registerHybrid({
+      makeNativeMethod("initHybrid", RNGestureHandlerComponentsRegistry::initHybrid),
+  });
+  
+  // This is a temporary solution that allows components exported by the screens
```
Comment (WoLewicki) [inline android/src/main/jni/RNGestureHandlerComponentsRegistry.cpp:61]: change all `screens` to `gesture-handler`

File: android/src/paper/java/com/swmansion/gesturehandler/ReactContextExtensions.kt [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package com.swmansion.gesturehandler
+
+import com.facebook.react.bridge.ReactContext
+import com.facebook.react.uimanager.UIManagerModule
+import com.facebook.react.uimanager.events.Event
+
+fun ReactContext.dispatchEvent(event: Event<*>) {
+    this.getNativeModule(UIManagerModule::class.java)!!.eventDispatcher.dispatchEvent(event)
```
Comment (WoLewicki) [inline android/src/paper/java/com/swmansion/gesturehandler/ReactContextExtensions.kt:8]: Shouldn't `let` be used here to be `null-safe`?

File: ios/RNGestureHandlerButtonComponentView.mm [hunk 9]
```diff
@@ -0,0 +1,48 @@
+#ifdef RN_FABRIC_ENABLED
```
Comment (WoLewicki) [inline ios/RNGestureHandlerButtonComponentView.mm:1]: It is added to reduce the final code size on paper? If so, should we do the same but inverted for Fabric code?

File: src/utils.ts [hunk 10]
```diff
@@ -26,14 +26,20 @@ export function withPrevAndCurrent<T, Transformed>(
...
 
-export function hasProperty(object: Record<string, unknown>, key: string) {
+// eslint-disable-next-line @typescript-eslint/ban-types
+export function hasProperty(object: object, key: string) {
   return Object.prototype.hasOwnProperty.call(object, key);
 }
 
 export function isJestEnv(): boolean {
-  return !!process.env.JEST_WORKER_ID;
+  return hasProperty(global, 'process') && !!process.env.JEST_WORKER_ID;
```
Comment (WoLewicki) [inline src/utils.ts:35]: Is this change needed in this PR?

File: src/jestUtils.ts [hunk 11]
```diff
@@ -60,6 +59,22 @@ import {
 import { State } from './State';
 import { hasProperty, withPrevAndCurrent } from './utils';
 
+// load fireEvent conditionally, so RNGH may be used in setups without testing-library
```
Comment (WoLewicki) [inline src/jestUtils.ts:62]: Is this change needed in this PR?

File: src/utils.ts (hunk 10 above)
Comment (WoLewicki) [inline src/utils.ts:35]: Is this change needed in this PR?

File: src/jestUtils.ts (hunk 11 above)
Comment (WoLewicki) [inline src/jestUtils.ts:62]: Is this change needed in this PR?

File: src/getShadowNodeFromRef.web.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+// Used by GestureDetector (unsupported on web at the moment) to check whether the
```
Comment (WoLewicki) [inline src/getShadowNodeFromRef.web.ts:1]: I would add this comment to non-web version too.

File: example/ios/Podfile [hunk 13]
```diff
@@ -6,7 +6,11 @@ platform :ios, '11.0'
 target 'Example' do
   config = use_native_modules!
 
-  use_react_native!(:path => config["reactNativePath"])
+  use_react_native!(
```
Comment (WoLewicki) [inline example/ios/Podfile:9]: Are all those changes of `example` project necessary in this PR? They seem to not concern it.

**Review by WoLewicki — COMMENTED** [review]
Very good job, keep it going! I left some comments, please answer them :tada:

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt (hunk 5 above)
Comment (j-piasecki) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt:545]: Events sent by `onHandlerUpdate` are handled by the `onUpdate` callback of a gesture.

File: android/src/main/AndroidManifest.xml [hunk 14]
```diff
@@ -1,3 +1,3 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.swmansion.gesturehandler.react">
+    package="com.swmansion.gesturehandler">
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply android/src/main/AndroidManifest.xml:2]: IIRC it's due to a bug in codegen where it ignores the set package name. Because of that the `RNGestureHandlerPackage` was moved from `com.swmansion.gesturehandler.react` to `com.swmansion.gesturehandler`. This change makes it work on Paper.

File: android/lib/src/main/java/com/swmansion/gesturehandler/GestureHandler.kt [hunk 15]
```diff
@@ -34,7 +34,7 @@ open class GestureHandler<ConcreteGestureHandlerT : GestureHandler<ConcreteGestu
     private set
   var isEnabled = true
     private set
-  var usesDeviceEvents = false
+  var actionType = 0
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply android/lib/src/main/java/com/swmansion/gesturehandler/GestureHandler.kt:37]: On Paper, we used direct events for Animated & Reanimated and device events for JS callbacks. On Fabric, we needed more granular control over them:
- events for Reanimated (for now only on Paper using direct events, will need special treatment on Fabric)
- events for Animated with nativeDriver=true (direct events)
- events for Animated with nativeDriver=false (direct events on Paper, device events on Fabric)
- events for JS callbacks (device events)

File: android/src/paper/java/com/swmansion/gesturehandler/ReactContextExtensions.kt (hunk 8 above)
Comment (j-piasecki) [reply android/src/paper/java/com/swmansion/gesturehandler/ReactContextExtensions.kt:8]: Should we be concerned about that? `UIManagerModule` shouldn't be `null` if everything is working correctly, no? But if that's the case, wouldn't `?.` work as well?

File: src/getShadowNodeFromRef.web.ts (hunk 12 above)
Comment (j-piasecki) [reply src/getShadowNodeFromRef.web.ts:1]: Done in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/a94c858205b10551ee7e2383f0bcf55d3d73f81f.

File: ios/RNGestureHandlerButtonComponentView.mm (hunk 9 above)
Comment (j-piasecki) [reply ios/RNGestureHandlerButtonComponentView.mm:1]: It implements `RCTRNGestureHandlerButtonViewProtocol` from codegen. This way it doesn't interfere with compilation on Paper where it's not used.

File: src/jestUtils.ts (hunk 11 above)
Comment (j-piasecki) [reply src/jestUtils.ts]: Not really, removed in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/af2e38b5d6a3996362bcbd9065fb8f54b2bdb894 in favor of https://github.com/software-mansion/react-native-gesture-handler/pull/1895.

Comment (j-piasecki) [reply src/jestUtils.ts]: Not really, removed in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/af2e38b5d6a3996362bcbd9065fb8f54b2bdb894 in favor of https://github.com/software-mansion/react-native-gesture-handler/pull/1895.

File: src/utils.ts (hunk 10 above)
Comment (j-piasecki) [reply src/utils.ts]: No, removed in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/1f7023072ff5d196ce6c479b44f67b6097e4fae8 in favor of https://github.com/software-mansion/react-native-gesture-handler/pull/1896.

Comment (j-piasecki) [reply src/utils.ts]: No, removed in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/1f7023072ff5d196ce6c479b44f67b6097e4fae8 in favor of https://github.com/software-mansion/react-native-gesture-handler/pull/1896.

File: FabricExample/src/ComponentsScreen.tsx [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+  ScrollView,
+  Switch,
+  TextInput,
+  TouchableNativeFeedback,
+  TouchableOpacity,
+} from 'react-native-gesture-handler';
+import { StyleSheet, Text, View } from 'react-native';
+
+import { COLORS } from './colors';
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply FabricExample/src/ComponentsScreen.tsx]: Done in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/850d510e768f3d06c92b65f42e7d42684066c1e6.

File: FabricExample/src/FinalScreen.tsx [hunk 17]
```diff
@@ -0,0 +1,55 @@
+import { Image, StyleSheet, Text, View } from 'react-native';
+
+import { COLORS } from './colors';
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply FabricExample/src/FinalScreen.tsx]: Done in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/850d510e768f3d06c92b65f42e7d42684066c1e6.

File: FabricExample/src/GestureCompositionScreen.tsx [hunk 18]
```diff
@@ -0,0 +1,120 @@
+import { Gesture, GestureDetector } from 'react-native-gesture-handler';
+import { StyleSheet, Text, View } from 'react-native';
+
+import { COLORS } from './colors';
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply FabricExample/src/GestureCompositionScreen.tsx]: Done in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/850d510e768f3d06c92b65f42e7d42684066c1e6.

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt (hunk 6 above)
Comment (tomekzaw) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt:554]: **1) Reanimated**

In the old architecture, Reanimated overwrites RCTEventDispatcher with REAEventDispatcher to intercept all direct events (see [here](https://github.com/software-mansion/react-native-reanimated/blob/a870aa28092322b627695f8cf2ea0dce4db34a53/ios/REAEventDispatcher.m#L8-L12)). This allows Gesture Handler to integrate with Reanimated without defining a hard dependency, simply by sending regular direct events.

In the new architecture, the event system was mainly ported to C++ so we can no longer overwrite RCTEventDispatcher. Instead, we need to dispatch events directly to Reanimated internals.

One more thing to note here is that in Reanimated for Paper there is no feedback if Reanimated did or did not handle specific event, so effectively they are handled twice (once by Reanimated and once by React Native). We plan to address this issue simultaneously with the adoption of Fabric.

**2) Animated.event with useNativeDriver: true**

Currently, `NativeAnimated[Turbo]Module` is a `RCTEventDispatcherObserver` (see [here](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/CoreModules/RCTEventDispatcher.mm#L117-L119)). Basically, it listens to all events going through `RCTEventDispatcher` but cannot prevent them from being sent to JS side. That's why in such case, similarly as for Reanimated, we would like to invoke NativeAnimated[Turbo]Module directly instead of flooding the app with unnecessary events.

**3) Direct vs. device events**

On Paper, the old Gesture Handler API (e.g. `PanGestureHandler`) uses direct events in order to invoke JS callbacks from the native side. When designing API for Gesture Handler 2.0 it turned out to be a limitation, so the new API (i.e. `GestureDetector`) uses device events instead.

With Fabric, at some point we decided to use only device events, both for old and new API. That's because I've already managed to get device events working in Gesture Handler but also it was kinda unclear how to send direct events from Obj-C. It looks like for this purpose we need to use C++ EventEmitter that we can access from the provided ShadowNode. However, currently Gesture Handler is set up as a NativeModule (in Java/Obj-C) so we would need to migrate it first to C++ and use JSI in order to pass ShadowNodes from JS.

Effectively, using direct or device events should be no difference for the library user as this seems to be an internal implementation detail of our library. Nevertheless, we will research into ways of sending direct events like we used to do on Paper.

File: android/src/main/jni/RNGestureHandlerComponentsRegistry.cpp (hunk 7 above)
Comment (tomekzaw) [reply android/src/main/jni/RNGestureHandlerComponentsRegistry.cpp]: Done in 786902cbba0fd1278f16de76130d53365fd08cb7

File: example/ios/Podfile (hunk 13 above)
Comment (tomekzaw) [reply example/ios/Podfile:9]: These changes restore the original Podfile from RN 0.67.2 template so we can build the app on our M1s. Will move them to a separate PR.

File: ios/RNGestureHandlerButtonComponentView.mm (hunk 9 above)
Comment (tomekzaw) [reply ios/RNGestureHandlerButtonComponentView.mm:1]: Another option would be to manually exclude these files in Podspec, either by using `*.{h,m}'` pattern for `source_files` or alternatively via `exclude_files` property, but we prefer the preprocessor conditionals.

File: RNGestureHandler.podspec [hunk 19]
```diff
@@ -1,5 +1,12 @@
 require "json"
 
+fabric_enabled = ENV['RCT_NEW_ARCH_ENABLED']
+
+# folly_version must match the version used in React Native
+# See folly_version in react-native/React/FBReactNativeSpec/FBReactNativeSpec.podspec
+folly_version = '2021.06.28.00-v2'
```
[the remark this answers is not in the document]
Comment (tomekzaw) [reply RNGestureHandler.podspec]: When it comes to configuration, we aimed to be compliant with react-native-screens, but your suggestion sounds like a good idea. We can also avoid overwriting `s.platforms`. Done in b745bbe68c06b1a789e6199ff41ad0619d153e9d. Would you accept an analogous PR to RNScreens?

File: FabricExample/android/settings.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomekzaw) [reply FabricExample/android/settings.gradle:8]: FabricExample is meant to be used only with Fabric enabled so that makes no difference at all here. Let's wait until RN teams publishes an official tutorial to see what's the recommended configuration.

File: android/src/main/AndroidManifest.xml (hunk 14 above)
[the remark this answers is not in the document]
Comment (tomekzaw) [reply android/src/main/AndroidManifest.xml:2]: To be precise, it looks like currently `codegenJavaPackageName` in `build.gradle` is never used. In particular, it does not seem to affect the package name of `RNGestureHandlerPackage` import when generating `PackageList.java` which is a part of auto-linking mechanism.

File: RNGestureHandler.podspec (hunk 19 above)
[the remark this answers is not in the document]
Comment (WoLewicki) [reply RNGestureHandler.podspec]: We are always open for PRs in RNScreens 😸

File: android/src/paper/java/com/swmansion/gesturehandler/ReactContextExtensions.kt (hunk 8 above)
Comment (WoLewicki) [reply android/src/paper/java/com/swmansion/gesturehandler/ReactContextExtensions.kt:8]: I have no knowledge if it should be present there, but I think it should be handled in a better way than `!!`, by e.g. throwing an exception with descriptive message or something like that.

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt (hunk 6 above)
Comment (WoLewicki) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt:554]: Maybe it would be nice to add this e.g. to the description of PR so it does not get lost?

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt (hunk 5 above)
Comment (WoLewicki) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt:545]: So maybe it would be good to extend this comment a bit so it is more descriptive?

File: android/lib/src/main/java/com/swmansion/gesturehandler/GestureHandler.kt (hunk 15 above)
[the remark this answers is not in the document]
Comment (WoLewicki) [reply android/lib/src/main/java/com/swmansion/gesturehandler/GestureHandler.kt:37]: Since @tomekzaw gave more detailed description of this below, I think we can resolve this.

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt (hunk 5 above)
Comment (j-piasecki) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt]: Good point, done in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/e51489fd2dc62a5e88f94fe0017bfec99081fc8a.

File: FabricExample/ios/FabricExample.xcodeproj/project.pbxproj [hunk 20]
```diff
@@ -566,7 +566,7 @@
 				COPY_PHASE_STRIP = NO;
 				ENABLE_STRICT_OBJC_MSGSEND = YES;
 				ENABLE_TESTABILITY = YES;
-				"EXCLUDED_ARCHS[sdk=iphonesimulator*]" = "";
+				"EXCLUDED_ARCHS[sdk=iphonesimulator*]" = i386;
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply FabricExample/ios/FabricExample.xcodeproj/project.pbxproj]: I believe it was changes because of problems with building the app on M1, but it seems to be working correctly without it. Removed in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/5fee6b1760d9c9f060f45b26149ca335a36fc8f7.

File: FabricExample/ios/FabricExample/AppDelegate.mm [hunk 21]
```diff
@@ -4,6 +4,8 @@
 #import <React/RCTBridge.h>
 #import <React/RCTBundleURLProvider.h>
 #import <React/RCTRootView.h>
+#import "MBFingerTipWindow.h"
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply FabricExample/ios/FabricExample/AppDelegate.mm:6]: For now I've removed `MBFingerTipWindow` entirely as it was crashing the app in some cases. Will look into this in the future.

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerButtonViewManager.kt (hunk 4 above)
Comment (j-piasecki) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerButtonViewManager.kt:351]: Yes, `android.R` was used before migration to Kotlin (https://github.com/software-mansion/react-native-gesture-handler/commit/48468306b48207b545b47f976dcdb9f69e43e546#diff-e614cdf726640239cd9b5e551cfc1c50eb1d6f078b8c6884bfa4fb321ed82e4dL209), I'm not sure why it was changed then.

File: android/src/paper/java/com/swmansion/gesturehandler/ReactContextExtensions.kt (hunk 8 above)
Comment (j-piasecki) [reply android/src/paper/java/com/swmansion/gesturehandler/ReactContextExtensions.kt]: I wrapped it with a try-catch block in https://github.com/software-mansion/react-native-gesture-handler/pull/1893/commits/589e641049566b13c7aafc72d9a7fda4bec0306d.

File: android/src/main/java/com/swmansion/gesturehandler/RNGestureHandlerPackage.kt [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+import com.facebook.soloader.SoLoader
+import com.swmansion.gesturehandler.BuildConfig
+import com.swmansion.gesturehandler.react.RNGestureHandlerModule
+import com.swmansion.gesturehandler.react.RNGestureHandlerRootViewManager
+import com.swmansion.gesturehandler.react.RNGestureHandlerButtonViewManager
+
+class RNGestureHandlerPackage : ReactPackage {
+  override fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> {
+    if (BuildConfig.IS_NEW_ARCHITECTURE_ENABLED) {
+      // For Fabric, we load c++ native library here, this triggers screen's Fabric
```
Comment (kmagiera) [inline android/src/main/java/com/swmansion/gesturehandler/RNGestureHandlerPackage.kt:16]: "screens's"?

File: android/src/main/jni/OnLoad.cpp [hunk 23]
```diff
@@ -0,0 +1,12 @@
+#include <fbjni/fbjni.h>
+
+#include "RNGestureHandlerComponentsRegistry.h"
+
+JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM *vm, void *) {
+  return facebook::jni::initialize(vm, [] {
+    // TODO: dvacca ramanpreet unify this with the way
```
Comment (kmagiera) [inline android/src/main/jni/OnLoad.cpp:7]: why?

File: ios/RNGestureHandler.h [hunk 24]
```diff
@@ -60,7 +54,7 @@ if (value != nil) { recognizer.prop = [RCTConvert type:value]; }\
 @property (nonatomic, readonly, nullable) UIGestureRecognizer *recognizer;
 @property (nonatomic, readonly, nullable) RNGestureHandlerPointerTracker *pointerTracker;
 @property (nonatomic) BOOL enabled;
-@property (nonatomic) BOOL usesDeviceEvents;
+@property (nonatomic, nonnull) NSNumber *actionType;
```
Comment (kmagiera) [inline ios/RNGestureHandler.h:57]: should be of `RNGestureHandlerActionType` type no?

File: ios/RNGestureHandlerActionType.h [hunk 25]
```diff
@@ -0,0 +1,8 @@
+#import <Foundation/Foundation.h>
+
+typedef NS_ENUM(NSInteger, RNGestureHandlerActionType) {
```
Comment (kmagiera) [inline ios/RNGestureHandlerActionType.h:3]: I think it'd be ok to move it to GestureHandler.h

File: ios/RNGestureHandlerManager.mm [hunk 26]
```diff
@@ -196,24 +210,82 @@ - (void)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer
 
 #pragma mark Events
 
-- (void)sendTouchEvent:(RNGestureHandlerEvent *)event
+- (void)sendStateChangeEvent:(RNGestureHandlerStateChange *)event withActionType:(nonnull NSNumber *)actionType
 {
-    [_eventDispatcher sendEvent:event];
+    switch ([actionType integerValue]) {
```
Comment (kmagiera) [inline ios/RNGestureHandlerManager.mm:215]: should just be `actionType` here and there is no need to use integerValue (I think)

File: android/src/main/java/com/swmansion/gesturehandler/RNGestureHandlerPackage.kt (hunk 22 above)
Comment (tomekzaw) [reply android/src/main/java/com/swmansion/gesturehandler/RNGestureHandlerPackage.kt]: Fixed in fe98498f3c9cdffd72a3b5803f82fe328bc33510

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt [hunk 27]
```diff
@@ -409,6 +417,21 @@ class RNGestureHandlerModule(reactContext: ReactApplicationContext?)
     }
   }
 
+  @ReactMethod(isBlockingSynchronousMethod = true)
+  fun install(): Boolean {
+    return try {
+      System.loadLibrary("rngesturehandler_modules")
```
[the remark this answers is not in the document]
Comment (tomekzaw) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt]: Done in f92834dc797947b31b89139ad4e6553443b7c184

File: android/src/main/jni/OnLoad.cpp (hunk 23 above)
Comment (tomekzaw) [reply android/src/main/jni/OnLoad.cpp:7]: Removed in 3f76094bf94748de6851c4cdde1db9674009024d

File: ios/RNGestureHandlerActionType.h (hunk 25 above)
Comment (tomekzaw) [reply ios/RNGestureHandlerActionType.h:3]: We did it analogously to `RNGestureHandlerState.h`.

File: ios/RNGestureHandlerManager.mm (hunk 26 above)
Comment (tomekzaw) [reply ios/RNGestureHandlerManager.mm]: Done in 3c381aa2075734797c5a997d6708843494d93dce

File: example/ios/Podfile (hunk 13 above)
Comment (tomekzaw) [reply example/ios/Podfile]: Done in 58f752099ae1bbf53377258b61db4e9a8bd42a26, in favor of #1903

**Review by kmagiera — APPROVED** [review]

File: android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt (hunk 6 above)
Comment (rashmi-priya) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt:555]: > However, currently Gesture Handler is set up as a NativeModule (in Java/Obj-C) so we would need to migrate it first to C++ and use JSI in order to pass ShadowNodes from JS.

> Why is Gesture Handler NativeModule (in Java/Obj-C) was not changed to TurboModule?

Comment (tomekzaw) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt:555]: Hey @rashmi-priya! The reason why we didn't migrate RNGH NativeModule to a TurboModule is the fact that NativeModules still work on the new architecture. The goal of the migration was to add Fabric support to RNGH 2.x, so the package itself supports all versions of RN 0.63.0+ as well as the old and the new architecture.

TurboModules are backward-compatible (meaning they work as NativeModules on the old architecture) but we couldn't just rewrite NativeModule into a TurboModule because it would not compile on older versions of RN (e.g. 0.64).

Another option was to conditionally enable NativeModule on Paper and TurboModule on Fabric. However, in this variant we would duplicate code and the secondary goal for this PR was to introduce as few changes as possible.

We decided to leave NativeModule as it is and address this issue in the future. Maybe now it's a good time to do so. 😄 

I hope this answers your questions.

Comment (rashmi-priya) [reply android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt:555]: Thanks for the clarification.
